---
title: "Mini procjet 2: Multiple tests"
author: "Mateus Auza Cruz"
format: live-html
toc: True
execute:
  #echo: false
  error: true
  warning: false
  message: false
  
code-fold: true
knitr:
  opts_chunk: 
    tidy: styler
    message: false
    
embed-resources: true
df-print: kable
---

```{r}
#| include: false
#library(tidyverse)
#library(magrittr)
#library(gt)
#library(janitor)
#library(ggrepel)
#library(factoextra)
#library(caret)
library(readr)
library(SummarizedExperiment)
library(ggplot2)

# customizations
#theme_set(theme_light()) # or any theme that you like
```

# Introduction

This project analyzes a gene expression dataset stored as a SummarizedExperiment object in R. The dataset contains microarray-based gene expression measurements from leukemia patients.

The objective of this descriptive analysis is to:

-   Explore the structure of the dataset

-   Determine the number of genes and samples

-   Identify the experimental groups

-   Summarize the distribution of samples across cancer types

The dataset compares two types of leukemia:

1.  **ALL** – Acute Lymphoblastic Leukemia

2.  **AML** – Acute Myeloid Leukemia

The working directory was set to the location of the dataset file, and required Bioconductor packages were installed.

```{r}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("SummarizedExperiment")
gollub=readr::read_rds("Gollub_SE.rds")
```

The dataset was loaded as a `SummarizedExperiment` object, a standard Bioconductor structure used to store omics data including:

-   Expression matrix (assays)

-   Sample metadata (colData)

-   Gene metadata (rowData)

# Descriptive analysis

From the object structure, we have the Expression Matrix which contains:

-   Number of genes (features): 3051

-   Number of samples: 38

-   Expression assay name: "expr"

-   Data type: Numeric matrix

```{r}
# Dimensions
expr_matrix = assay(gollub, "expr")
```
Thus, the expression matrix has *3051 genes $\times$ 38 samples*.

The colData slot contains clinical information about samples:

-   cancer_class (numeric coding)

-   cancer_type (factor variable)

To determine the number of experimental groups:
```{r}
group = colData(gollub)$cancer_type
table(group)
```

To explore global gene expression differences between the two leukemia types, we computed the average expression per gene within each group and calculated the difference:

```{r}
#| results: hold
mean_ALL = rowMeans(expr_matrix[, group == "ALL"])
mean_AML = rowMeans(expr_matrix[, group == "AML"])
mean_diff = mean_AML - mean_ALL

summary(mean_diff)
hist(mean_diff, breaks = 50)
```

The mean difference is close to zero, indicating no global shift in expression between AML and ALL. The distribution is symmetric, suggesting that roughly half of the genes are upregulated in each group.

Although most genes show small differences, the wide range (−2.16 to +2.89) indicates that a subset of genes exhibits strong differential expression.

Overall, this suggests that specific genes, rather than a global transcriptomic shift, distinguish AML from ALL.

```{r}
boxplot(
  colMeans(expr_matrix) ~ group,
  col = c("lightblue", "salmon"),
  main = "Average Expression per Sample",
  ylab = "Mean Expression"
)

```

The boxplot shows:

-   The ALL group has slightly lower average expression.

-   The AML group shows higher median expression and slightly larger variability (IQR).

-   The boxplot shows two potential outliers: one in the ALL group and one in the AML group.

```{r}
#| fig.width: 14
#| fig.height: 10
# PCA analysis - logical in hig dimmensional analysis
group <- as.factor(group)
pca = prcomp(t(expr_matrix), scale. = TRUE)

pca_df <- data.frame(
  PC1 = pca$x[,1],
  PC2 = pca$x[,2],
  group = group
)

ggplot(pca_df, aes(x = PC1, y = PC2, color = group)) +
  geom_point(size = 3) +
  theme_minimal() +
  labs(title = "PCA of Gene Expression")
```

The PCA plot shows a clear separation between AML and ALL samples primarily along PC1, indicating that the main source of variation in the dataset corresponds to leukemia subtype. Samples cluster tightly within groups, suggesting strong between-group differences relative to within-group variability. This confirms the presence of a strong biological signal distinguishing AML from ALL.

#### Checking Normality Assumption

We assess now whether the gene expression values are approximately normally distributed using Q–Q plots.

```{r}
par(mfrow=c(1,2))
qqnorm(mean_ALL)
qqline(mean_ALL, col="red")

qqnorm(mean_AML)
qqline(mean_AML, col="green")
```
The Q–Q plots indicate that the distributions are approximately normal in the center, with some deviation in the tails. Since the t-test is robust to moderate deviations from normality, particularly for moderate sample sizes, we proceed with a two-sample t-test.

Because we do not assume equal variances between groups, we use Welch’s t-test (`var.equal= F`).


# Test and computation of p-values

To formally test for differential expression, we perform gene-wise two-sample, two-sided t-tests.

For each gene \( g \), we test the null hypothesis:

$$
H_0^{(g)} : \mu_{ALL}^{(g)} = \mu_{AML}^{(g)}
$$

where  

$$
\mu_{ALL}^{(g)}
$$

is the population mean expression of gene \( g \) in the ALL group, and  

$$
\mu_{AML}^{(g)}
$$

is the population mean expression of gene \( g \) in the AML group.

We then examined the distribution of p-values:
```{r}
group <- as.factor(group)

p_values <- apply(expr_matrix, 1, function(gene_expr) {
  t.test(gene_expr ~ group, var.equal=F)$p.value
})

hist(
  p_values,
  breaks = 50,
  col = "skyblue",
  main = "Histogram of p-values",
  xlab = "p-value"
)
```

The histogram is strongly left-skewed, with many p-values close to 0, rather than following the expected Uniform(0,1) distribution under the global null hypothesis (mean ≈ 0.5).
The observed mean p-value (0.28) and enrichment of small p-values indicate that many genes are truly differentially expressed.

```{r}
#| results: hold
mean(p_values)
sum(p_values < 0.05)
```

We performed 3051 independent two-sample t-tests comparing ALL and AML samples. At significance level $\alpha$ = 0.05 (uncorrected), 1078 genes were found significant. However, under the global null hypothesis, approximately 153 false positives would be expected.

# Multiple testing correction

Because we performed 3051 simultaneous hypothesis tests, controlling for multiple comparisons is essential to reduce false discoveries.

We applied both:

-   FWER (Family-Wise Error Rate) methods

-   FDR (False Discovery Rate) methods

## FWER methods

FWER methods control the probability of making at least one false positive among all tests.

##### Bonferroni correction

The Bonferroni correction adjusts p-values by multiplying by the number of tests 
m.

```{r}
m= length(p_values)
p_bonf = p.adjust(p_values, method = "bonferroni")
sum(p_bonf < 0.05)
```

##### Sidak correction

```{r}
p_sidak <- 1 - (1 - p_values)^m
sum(p_sidak < 0.05)
```

## FDR family

FDR methods control the expected proportion of false discoveries among the rejected hypotheses, which is more appropriate for high-dimensional omics data.

##### Benjamini-Hochberg correction

```{r}
p_bh <- p.adjust(p_values, method = "BH")
sum(p_bh < 0.05)
```

## Comparison

```{r}
data.frame(
  Method = c("Bonferroni", "Sidak", "BH"),
  Significant_Genes = c(
    sum(p_bonf < 0.05),
    sum(p_sidak < 0.05),
    sum(p_bh < 0.05)
  )
)

```

Bonferroni and Šidák are very conservative, drastically reducing the number of discoveries. However, in high-dimensional genomic studies, such strict control may eliminate many biologically meaningful genes. Thus, we opt for the Benjamini–Hochberg FDR correction.

## Top Differentially Expressed Genes

The ten most statistically significant genes based on raw p-values are:

```{r}
gene_names <- rowData(gollub)$gene_name

top10_idx <- order(p_values)[1:10]
top10_genes <- gene_names[top10_idx]

data.frame(
  Gene = top10_genes,
  p_values= format(p_values[top10_idx], scientific = TRUE, digits = 3)
)

```

# Volcano plot

The volcano plots visualize log2 Fold Change vs −log10(p-value) (statistical significance).
We will plot the volcano plots of:
-   Raw p-values
-   Bonferroni corrected p-values
-   Benjamini-Hochberg corrected p-values

```{r}
# Calculate mean expression in each group
mean_expr <- apply(expr_matrix, 1, function(gene_expr) {
  tapply(gene_expr, group, mean)
})

# log2 fold change: AML / ALL (assuming ALL = 0, AML = 1)
log2FC <- log2(mean_expr[2, ] / mean_expr[1, ])
```

```{r}
volcano_df <- data.frame(
  gene = rowData(gollub)$gene_name,
  log2FC = log2FC,
  p_raw = p_values,
  p_bonf = p_bonf,
  p_sidak = p_sidak,
  p_bh = p_bh
)

plot_volcano <- function(df, p_col, title) {
  plot(
    df$log2FC,
    -log10(df[[p_col]]),
    pch = 19,
    col = ifelse(df[[p_col]] < 0.05, "red", "grey"),
    xlab = "log2 Fold Change",
    ylab = "-log10(p-value)",
    main = title
  )
  abline(h = -log10(0.05), col = "blue", lty = 2) # significance line
  abline(v = c(-1, 1), col = "darkgreen", lty = 2) # fold-change cutoff
}
volcano_df_clean <- volcano_df[!is.nan(volcano_df$log2FC), ]
```

```{r}
par(mfrow = c(1, 3)) # 3 plots side by side

plot_volcano(volcano_df_clean, "p_raw", "Raw p-values")
plot_volcano(volcano_df_clean, "p_bonf", "Bonferroni")
plot_volcano(volcano_df_clean, "p_bh", "Benjamini-Hochberg")
```

**Observations:**

Volcano plots show that raw p-values overestimate significance due to multiple testing, Bonferroni correction is overly conservative, and Benjamini–Hochberg provides a balanced approach by controlling false discovery rate while preserving biological signal.

**Interpretation: **

The volcano plots show a clear reduction of significant genes after multiple-testing correction. Raw p-values yield many apparent discoveries, but this likely includes numerous false positives. Bonferroni and Šidák corrections are highly stringent, retaining only the strongest signals. The Benjamini–Hochberg (BH) adjustment preserves a substantially larger set of significant genes while controlling the false discovery rate.

# Conclusion

Across 3051 tests, 1078 genes were significant at α = 0.05 (uncorrected), far exceeding the ~153 false positives expected under the global null, indicating genuine differential expression. After correction, Bonferroni and Šidák each identify 103 genes, whereas BH identifies 695 genes, reflecting greater sensitivity. Given the high-dimensional context, BH provides the best balance between discovery and error control. 

**Therefore, there is strong evidence that a robust set of genes can differentiate ALL and AML samples.**
